ur_robot_driver/src/ros/hardware_interface.cpp

```diff
@@ -628,6 +648,50 @@ void HardwareInterface::publishPose()
   }
 }
 
+void HardwareInterface::extractRobotStatus()
+{
+  using namespace rtde_interface;
+
+  robot_status_resource_.mode = robot_status_bits_[toUnderlying(UrRtdeRobotStatusBits::IS_TEACH_BUTTON_PRESSED)] ?
+                                    RobotMode::MANUAL :
+                                    RobotMode::AUTO;
+
+  robot_status_resource_.e_stopped = safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_EMERGENCY_STOPPED)] ?
+                                         TriState::TRUE :
+                                         TriState::FALSE;
+
+  // Note that this is true as soon as the drives are powered,
+  // even if the breakes are still closed
+  // which is in slight contrast to the comments in the
+  // message definition
+  robot_status_resource_.drives_powered =
+      robot_status_bits_[toUnderlying(UrRtdeRobotStatusBits::IS_POWER_ON)] ? TriState::TRUE : TriState::FALSE;
+
+  robot_status_resource_.motion_possible =
+      robot_mode_ == ur_dashboard_msgs::RobotMode::RUNNING ? TriState::TRUE : TriState::FALSE;
+
+  // I found no way to reliably get information if the robot is moving
+  robot_status_resource_.in_motion = TriState::UNKNOWN;
+
+  if (safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_PROTECTIVE_STOPPED)] ||
+      safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_ROBOT_EMERGENCY_STOPPED)] ||
+      safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_EMERGENCY_STOPPED)] ||
+      safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_VIOLATION)] ||
+      safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_FAULT)] ||
+      safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_STOPPED_DUE_TO_SAFETY)])
```

Comment:
This may be taking it too far, but: seeing as this set never changes, what about keeping a "mask" `bitset` around, initialised once and then used as:

```c++
if (safety_status_bits_ | in_error_bitset_)
   ...
```

`std::bitset` supports [the necessary operators](https://en.cppreference.com/w/cpp/utility/bitset/operator_logic2).

Reply:
I'll leave that up to @axelschroth It is fine like that from my side.

Reply:
We could ignore it in this PR, but seeing as this is run @ the controller's update rate (ie: 125 or 500 Hz), it's 7 operations vs 1. I'm not sure the compiler will be able to coalesce all the `bitset::test(..)` invocations into a single masking operation.

Reply:
true... And it should be fairly quick to implement :-)